csharp/client/Dh_NetClient/container/RowSequence.cs

```diff
@@ -0,0 +1,174 @@
+﻿//
+// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
+//
+using System;
+
+namespace Deephaven.Dh_NetClient;
+
+public abstract class RowSequence {
+  public static RowSequence CreateEmpty() => SequentialRowSequence.EmptyInstance;
+
+  public static RowSequence CreateSequential(Interval interval) {
+    return new SequentialRowSequence(interval);
+  }
+
+  public bool IsEmpty => Count == 0;
+  public abstract UInt64 Count { get; }
+
+  public abstract IEnumerable<Interval> Intervals { get; }
+
+  public IEnumerable<UInt64> Elements {
+    get {
+      foreach (var interval in Intervals) {
+        for (UInt64 i = interval.Begin; i != interval.End; ++i) {
+          yield return i;
+        }
+      }
+    }
+  }
+
+  public abstract RowSequence Take(UInt64 size);
+  public abstract RowSequence Drop(UInt64 size);
+}
+
+sealed class SequentialRowSequence(Interval interval) : RowSequence {
+  public static readonly SequentialRowSequence EmptyInstance = new(Interval.OfEmpty);
+
+  public override UInt64 Count => interval.Count;
+
+  public override IEnumerable<Interval> Intervals {
+    get {
+      yield return interval;
+    }
+  }
+
+  public override RowSequence Take(UInt64 size) {
+    var sizeToUse = Math.Min(size, Count);
+    return new SequentialRowSequence(interval with { End = interval.Begin + sizeToUse });
+  }
+
+  public override RowSequence Drop(UInt64 size) {
+    var sizeToUse = Math.Min(size, Count);
+    return new SequentialRowSequence(interval with { Begin = interval.Begin + sizeToUse });
+  }
+}
+
+sealed class BasicRowSequence : RowSequence {
+  public static BasicRowSequence Create(IEnumerable<Interval> intervals) {
+    var intervalsArray = intervals.ToArray();
+    UInt64 count = 0;
+    foreach (var e in intervalsArray) {
+      count += e.Count;
+    }
+
+    return new BasicRowSequence(intervalsArray, 0, 0, count);
+  }
+
+  private readonly Interval[] _intervals;
+  private readonly int _startIndex = 0;
+  private readonly UInt64 _startOffset = 0;
+  public override UInt64 Count { get; }
+
+  private BasicRowSequence(Interval[] intervals, int startIndex, UInt64 startOffset, UInt64 count) {
+    _intervals = intervals;
+    _startIndex = startIndex;
+    _startOffset = startOffset;
+    Count = count;
+  }
+
+  public override RowSequence Take(UInt64 count) =>
+    new BasicRowSequence(_intervals, _startIndex, _startOffset, count);
+
+  public override RowSequence Drop(UInt64 count) {
+    return DropHelper(count).Last().Item2!;

```

Comment:
Can you add a comment here? I think what is happening is an iterator up to `count` is exhausted then the final returned. But I dunno the final `!` meaning.

Reply:
I have added more verbose comments. Let me know if that explains it somewhat. It *is* a bit strange.

`!` is the operator that tells the compiler "I know you think this value *could* be null but I'm asserting to you that it's not.

You might ask whether this can break type safety, but (unfortunately, no one checked with me), nullness annotations are advisory not compulsory and it doesn't break anything (other than NullReferenceExceptions) if you get it wrong.

Reply:
comments helped a lot, thanks!